dbms/src/Flash/Coprocessor/CHBlockChunkCodecV1.h

```diff
@@ -0,0 +1,292 @@
+                    RUNTIME_ASSERT(rows == elem->size());
+                    bytes += elem->byteSize();
+                }
+            }
+        }
+    }
+    static const ColumnPtr & toColumnPtr(const Columns & c, size_t index)
+    {
+        return c[index];
+    }
+    static ColumnPtr toColumnPtr(MutableColumns && c, size_t index)
+    {
+        return std::move(c[index]);
+    }
+    static ColumnPtr toColumnPtr(const MutableColumns & c, size_t index)
+    {
+        return c[index]->getPtr();
+    }
+    static const ColumnPtr & toColumnPtr(const Block & block, size_t index)
+    {
+        return block.getByPosition(index).column;
+    }
+
+    template <typename ColumnsHolder>
+    constexpr static bool isBlockType()
+    {
+        return std::is_same_v<std::remove_const_t<std::remove_reference_t<ColumnsHolder>>, Block>;
+    }
+
+    template <typename ColumnsHolder>
+    size_t getRows(ColumnsHolder && columns_holder)
+    {
+        if constexpr (isBlockType<ColumnsHolder>())
+        {
+            const Block & block = columns_holder;
+            size_t rows = block.rows();
+            return rows;
+        }
+        else
+        {
+            if unlikely (!columns_holder.front())
+                return 0;
+            size_t rows = columns_holder.front()->size();
+            return rows;
+        }
+    }
+
+    template <typename ColumnsHolder>
+    void encodeColumnImpl(ColumnsHolder && columns_holder, WriteBuffer * ostr_ptr)
+    {
+        size_t rows = getRows(std::forward<ColumnsHolder>(columns_holder));
+        if (!rows)
+            return;
+
+        // Encode row count for next columns
+        writeVarUInt(rows, *ostr_ptr);
+
+        // Encode columns data
+        for (size_t col_index = 0; col_index < header.columns(); ++col_index)
+        {
+            auto && col_type_name = header.getByPosition(col_index);
+            auto && column_ptr = toColumnPtr(std::forward<ColumnsHolder>(columns_holder), col_index);
+            WriteColumnData(*col_type_name.type, column_ptr, *ostr_ptr, 0, 0);
+        }
+
+        encoded_rows += rows;
+    }
+    void encodeColumn(const MutableColumns & columns, WriteBuffer * ostr_ptr)
+    {
+        return encodeColumnImpl(columns, ostr_ptr);
+    }
+    void encodeColumn(const Columns & columns, WriteBuffer * ostr_ptr)
+    {
+        return encodeColumnImpl(columns, ostr_ptr);
+    }
+    void encodeColumn(const std::vector<MutableColumns> & batch_columns, WriteBuffer * ostr_ptr)
+    {
+        for (auto && batch : batch_columns)
+        {
+            encodeColumnImpl(batch, ostr_ptr);
+        }
+    }
+    void encodeColumn(std::vector<MutableColumns> && batch_columns, WriteBuffer * ostr_ptr)
+    {
+        for (auto && batch : batch_columns)
+        {
+            encodeColumnImpl(std::move(batch), ostr_ptr);
+        }
+    }
+    void encodeColumn(const Block & block, WriteBuffer * ostr_ptr)
+    {
+        assert(&block == &header);
```

Comment:
I think `header` should 0 rows, does this mean the `block` does not have any data?

Reply:
I have refine this part. To encode a single block, it can be like
```
                auto codec = CHBlockChunkCodecV1{block, true}; // block to be encoded
                auto && str = codec.encode(CompressionMethod::LZ4);
                assert(!str.empty());
```